### ISSUE TITLE
Improve string concatenation with non-const chars to use spans

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#66827]
Today for string concatenation, Roslyn will emit a ToString call on value type arguments in order to avoid also paying for boxing, e.g. this:
```C#
public static string M1(string s, char c) => s + c + s;
```
is compiled to the equivalent of:
```C#
public static string M1(string s, char c) => string.Concat(s, c.ToString(), s);
```
For chars in particular, however, it could now do better. .NET 7 introduced a new safe constructor on `ReadOnlySpan<T>` that lets you wrap a span around a variable, e.g.
```C#
new ReadOnlySpan<char>(in c)
```
and `string.Concat` has overloads that accept 2, 3, and 4 `ReadOnlySpan<char>` arguments.  As such, this could instead be lowered to:
```C#
public static string M1(string s, char c) => string.Concat(s, new ReadOnlySpan<char>(in c), s);
```
when targeting a platform that has that ctor, avoiding the extra string allocation.

Note that the compiler already handles the case of upgrading const chars to be const strings, so this is only applicable to non-const chars.

**Comment by AlekseyTs** [comment]
I am curious where and whether should we draw a line for the optimization.

For example:
```
var a = "a";
var b = "b";
var c = "c";
var d = "d";
var e = 'e';
                                                  
_ = a + b + c + (d + e); 
``` 

Is it better to go with the old code gen and allocate string for char, or is it better to take advantage of "span" concat and allocate bigger intermediate string because of that?
Is the answer going to be different if we add more string operands to the picture? 

Opinions?

**Comment by stephentoub** [comment]
> I am curious where and whether should we draw a line for the optimization.

From my perspective, we should draw the line at whether there's an overload of string.Concat that can do the operation efficiently, defined as the caller not needing to allocate anything.

> Is it better to go with the old code gen and allocate string for char, or is it better to take advantage of "span" concat and allocate bigger intermediate string because of that?

You're asking is it better to generate:
```C#
string.Concat("a", "b", "c", string.Concat("d", new ReadOnlySpan<char>(in e)));
```
or
```C#
string.Concat(new string[] { "a", "b", "c", "d", e.ToString() }); // params array
```
?

With these _exact_ inputs, I'd expect the former to be better: you have only one intermediate allocation rather than two, and the intermediate string will be the same size in both cases, because the extra "d" won't increase the size of the string object due to padding. However, the equation changes if these inputs become variable, because the former will be allocating a string object that includes all of d, so as d grows, so too does that intermediate allocation, whereas in the latter, the intermediate string only includes the e character and not the d string... at some length of d, it starts to be more expensive than the extra string[].

**Comment by AlekseyTs** [comment]
@stephentoub 

It looks like you are assuming that we somehow know content (at least the length) for the strings involved. We don't. Therefore, how the precise decision algorithm would look like, according to you?

To be honest, I do not know how to implement: "we should draw the line at whether there's an overload of string.Concat that can do the operation efficiently, defined as the caller not needing to allocate anything." Mostly because this criteria is too vague. My assumption is that `String.Concat(ReadOnlySpan, ReadOnlySpan)` allocates in a general case. Therefore, both code gen strategies allocate, from my perspective.      

Here is another example (numbers at the end of arguments do not indicate anything about content of the strings, we do not know what it is going to be at runtime):
```
 string.Concat(new string[] { string1, string2, string3, string4, string5, char1 }); // params array
```
vs.
```
string.Concat(new string[] { string1, string2, string3, String4, string.Concat(string5, new ReadOnlySpan<char>(in char1) }); // params array
```

Which approach is more preferable?

**Comment by stephentoub** [comment]
> It looks like you are assuming that we somehow know content (at least the length) for the strings involved

I'm not assuming anything. You asked which approach is better, and I highlighted that you can't know unless you know all the inputs.

> My assumption is that String.Concat(ReadOnlySpan, ReadOnlySpan) allocates in a general case

The _caller_ doesn't allocate, whereas when using the params overload, the caller does allocate, and when using ToString on an argument, the caller does allocate.

> Which approach is more preferable?

That first example won't compile, as you can't put a char into a string array.

**Comment by AlekseyTs** [comment]
> That first example won't compile (`string.Concat(new string[] { string1, string2, string3, string4, string5, char1 }); // params array`), as you can't put a char into a string array.

My mistake. Corrected in place.

**Comment by AlekseyTs** [comment]
Sorry, I am still confused. Does ```String.Concat(ReadOnlySpan, ReadOnlySpan)``` allocates a new string according to you, or not? If it does, is this something that the algorithm should take into account?

I am still hoping that you can provide an algorithm for the decision process, which can be applied to a arbitrary set of inputs. Some of the inputs are strings and some of the inputs are chars.

**Comment by stephentoub** [comment]
> Does String.Concat(ReadOnlySpan, ReadOnlySpan) allocates a new string according to you, or not?

It allocates the single string to be returned from the Concat, nothing else. Every overload of string.Concat does that, it's the purpose of Concat. The only time string.Concat won't allocate that result string is when no concatenation needs to be performed, e.g. because all of the inputs are empty.

> I am still hoping that you can provide an algorithm for the decision process, which can be applied to an arbitrary set of inputs. Some of the inputs are strings and some of the inputs are chars.

If there's a concatenation of strings and chars, and if there's an available overload of string.Concat that takes the corresponding number of `ReadOnlySpan<char>`s (which today means 2, 3, or 4), use that overload by converting each of the arguments to a span.

**Comment by AlekseyTs** [comment]
> If there's a concatenation of strings and chars, and if there's an available overload of string.Concat that takes the corresponding number of ReadOnlySpan<char>s (which today means 2, 3, or 4), use that overload by converting each of the arguments to a span.

Should I interpret that, if there are 5 inputs, you do not expect the codegen to change?

**Comment by stephentoub** [comment]
> Should I interpret that like, if there are 5 inputs, you do not expect the codegen to change?

Correct.

**Comment by AlekseyTs** [comment]
Thanks!

**Comment by stephentoub** [comment]
(We could subsequently have a discussion about using interpolated string handlers for lowering string concatenation in some cases, but that's a separate conversation.)

**Comment by DoctorKrolic** [comment]
>> Should I interpret that like, if there are 5 inputs, you do not expect the codegen to change?
>
> Correct.

I just want to point out that according to my measurements if we have e.g. 5 arguments to concat, where 1 of them is char it is still benifitial to concat it in 2 rounds: first use span-based concat of 4 and then concat the remaining one:
```
BenchmarkDotNet v0.13.12, Windows 11 (10.0.22621.3007/22H2/2022Update/SunValley2)
AMD Ryzen 7 5800X, 1 CPU, 16 logical and 8 physical cores
.NET SDK 8.0.101
  [Host]     : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2
  DefaultJob : .NET 8.0.1 (8.0.123.58001), X64 RyuJIT AVX2


| Method        | Mean     | Error    | StdDev   | Gen0   | Allocated |
|-------------- |---------:|---------:|---------:|-------:|----------:|
| ConcatStrings | 29.96 ns | 0.369 ns | 0.345 ns | 0.0072 |     120 B |
| ConcatSpans   | 17.18 ns | 0.103 ns | 0.096 ns | 0.0038 |      64 B |
```
```cs
using BenchmarkDotNet.Attributes;
using BenchmarkDotNet.Running;

BenchmarkRunner.Run<Benchmarks>();

[MemoryDiagnoser]
public class Benchmarks
{
    [Benchmark]
    public string ConcatStrings()
        => M1("s", 'c');

    [Benchmark]
    public string ConcatSpans()
        => M2("s", 'c');

    private static string M1(string s, char c)
    {
        return string.Concat(s, c.ToString(), s, s, s);
    }

    private static string M2(string s, char c)
    {
        return string.Concat(string.Concat(s, new ReadOnlySpan<char>(in c), s, s), s);
    }
}
```
Do we preffer concat with params array because this trend doesn't work for longer inputs/more concat parameters or at this point we preffer IL size over runtime performance?

**Comment by AlekseyTs** [comment]
@DoctorKrolic It sounds like you might be proposing a different algorithm. Could you summarize it? And perhaps include an explanation for why you think it is worth using it instead.

**Comment by DoctorKrolic** [comment]
> It sounds like you might be proposing a different algorithm

Sorry, I'm not sure what do you mean by that. The reason I clarified about 5+ concat args case is because we will probably need to do additional steps in our implementation to ensure we pick params overload in all cases. And because when I was first reading through lines I somehow missed Stephen's original explanation. Other than that the high-level goal is clear: when we have 3 or 4 concat args, 1 of which is non-constant char, pick span-based concat overaload when possible. Otherwise keep current codegen. The question _how_ we do it is an implementation detail of the PR, thus it should be discussed there
